src/v/cloud_storage_clients/abs_client.cc

```diff
@@ -731,20 +739,34 @@ abs_client::delete_object(
                  }),
                  key)
           .then([&name, &key](const ret_t& result) {
-              // ABS returns a 404 for attempts to delete a blob that doesn't
-              // exist. The remote doesn't expect this, so we map 404s to a
-              // successful response.
-              if (!result && result.error() == error_outcome::key_not_found) {
-                  vlog(
-                    abs_log.debug,
-                    "Object to be deleted was not found in cloud storage: "
-                    "object={}, bucket={}. Ignoring ...",
-                    name,
-                    key);
-                  return ss::make_ready_future<ret_t>(no_response{});
-              } else {
-                  return ss::make_ready_future<ret_t>(result);
+              if (!result) {
+                  if (result.error() == error_outcome::key_not_found) {
+                      // ABS returns a 404 for attempts to delete a blob that
+                      // doesn't exist. The remote doesn't expect this, so we
+                      // map 404s to a successful response.
+                      vlog(
+                        abs_log.debug,
+                        "Object to be deleted was not found in cloud storage: "
+                        "object={}, bucket={}. Ignoring ...",
+                        name,
+                        key);
+                      return ss::make_ready_future<ret_t>(no_response{});
+                  } else if (
+                    result.error() == error_outcome::operation_not_supported) {
+                      // ABS does not allow for deletion of directories, as they
+                      // are treated like a flat namespace. The "folder" is
+                      // "removed" when all blobs inside of it are deleted.
+                      // Map this to a successful response.
```

Comment:
When using the ADLS client, it seems we explicitly attempt to delete directories. Does that still make sense, based on what you've found?

And just making sure I understand the original problem, was it that we
1. didn't have HNS enabled, so were using the regular non-ADLS client, and
2. had a bug in the purger that caused us to try deleting directories, so that
3. the regular ABS client tripped up here with this new error code

Is that right? And if 1 or 2 weren't true we wouldn't have seen this?

Reply:
You are right, maybe I should have a clarifying sentence here- ABS does not allow for the deletion of directories when ADLS/HNS is not enabled, which is where this comment is in the code path. I can amend this to make it clearer.

And yes, I believe that sequence of events is what has bubbled up all these ABS related bugs. I believe we would have still seen the purger bug that attempted to treat a directory as a manifest regardless of 1.